### DISCUSSION TITLE
.NET images read from memory and Convert Commands

### DISCUSSION THREAD

**Discussion by EricBatlle** [dlemstra/Magick.NET#986] [General] [answered]
Hi @dlemstra , first of all, many thanks for your AWESOME library and your continuous maintenance! 

Now going to my question, I want to do some convert -auto-gamma and -normalize commands on to a bunch of images but doing it by CMD with something like:

$"magick convert *.png -normalize {path}" takes too long (nearly 5 seconds for 60 grayscale frames of 256x520 ) and I'm starting to think that this is caused to handling static images instead of memory-ones.

So my questions are: 
1. How can I work with memory images (byte[]) instead of static ones? I expected that this library with the `new ImageMagcik(byte[]) ` can do that, but I'm stuck trying to find how to apply convert to those virtual images...
2. There is a way with Magick.NET to achieve that?
3. Why takes so long to perform a -normalize filter?

In case it helps, I've tried to do my own normalize filter before reaching you, it's fast (really fast) but the results aren't the same as yours...
4.Why?

```
public class NormalizeFilter
{
	private static double maxGlobalPixelValue = 255;
	private static double minGlobalPixelValue = 0;
	private static double minThresholdPercentage = 0.02;
	private static double maxThresholdPercentage = 0.99;

	private double maxLocalPixelValue = 0;
	private double minLocalPixelValue = 255;

	private double normalizationFactor = 0;

	//Create filter using as normalization factor the frame passed
	public NormalizeFilter(byte[] framePixelsToCalculateNormalizationFactor, bool useThresholds = true)
	{
		normalizationFactor = useThresholds ? GetNormalizationFactor(framePixelsToCalculateNormalizationFactor) : GetNormalizationFactorWithoutThreshold(framePixelsToCalculateNormalizationFactor);
	}

	//Aply Normalize filter with normalizationFactor previously calculated
	public byte[] ApplyTo(byte[] framePixels)
	{
		byte[] framePixelsNormalizeds = new byte[framePixels.Length];

		double pixelUnnormalizedValue = 0;
		double pixelNormalizedValue = 0;
		for (int i = 0; i < framePixels.Length; i++)
		{
			pixelUnnormalizedValue = framePixels[i];
			pixelNormalizedValue = ((pixelUnnormalizedValue - minLocalPixelValue) * normalizationFactor) + minGlobalPixelValue;

			if (pixelNormalizedValue > maxGlobalPixelValue)
				pixelNormalizedValue = maxGlobalPixelValue;
			if (pixelNormalizedValue < minGlobalPixelValue)
				pixelNormalizedValue = minGlobalPixelValue;

			framePixelsNormalizeds[i] = (byte)pixelNormalizedValue;
		}

		return framePixelsNormalizeds;
	}

	private double GetNormalizationFactor(byte[] framePixels)
	{
		Array.Sort(framePixels);
		int minIndex = (int)(framePixels.Length * minThresholdPercentage);
		int maxIndex = (int)(framePixels.Length * maxThresholdPercentage);
		minLocalPixelValue = framePixels[minIndex];
		maxLocalPixelValue = framePixels[maxIndex];

		normalizationFactor = (maxGlobalPixelValue - minGlobalPixelValue) / (maxLocalPixelValue - minLocalPixelValue);

		return normalizationFactor;
	}

	//Instead of using thresholds to avoid outlayers simply use the maximum and minimum pixel values
	private double GetNormalizationFactorWithoutThreshold(byte[] framePixels)
	{
		for (int i = 0; i < framePixels.Length; i++)
		{
			if (framePixels[i] < minLocalPixelValue)
				minLocalPixelValue = framePixels[i];

			if (framePixels[i] > maxLocalPixelValue)
				maxLocalPixelValue = framePixels[i];
		}

		normalizationFactor = (maxGlobalPixelValue - minGlobalPixelValue) / (maxLocalPixelValue - minLocalPixelValue);

		return normalizationFactor;
	}
}
```

Again, many thanks for your time!

**Comment by dlemstra** [comment, marked as answer]
> How can I work with memory images (byte[]) instead of static ones? I expected that this library with the new ImageMagcik(byte[]) can do that, but I'm stuck trying to find how to apply convert to those virtual images...

I don't understand what you are asking me. If you want to read a single image you can specify the bytes of the image or the name of the file (and some other methods) as an argument to the `MagickImage` class.

> There is a way with Magick.NET to achieve that?

Still not sure what you are asking?

> Why takes so long to perform a -normalize filter?

The duration also depends on the version of ImageMagick that you are using. You are probably using a `Q16-HDRI` version on the command line. You might want to consider to use the `Q8` version of Magick.NET instead. And another thing that might help would be to use the `OpenMP` version.

And I suspect the duration is longer because we use double precision instead. More info about the implementation can be found here:

https://github.com/ImageMagick/ImageMagick/blob/8c4418952feafa9e5699404005d528a838db046a/MagickCore/enhance.c#L4119

And here:

https://github.com/ImageMagick/ImageMagick/blob/8c4418952feafa9e5699404005d528a838db046a/MagickCore/enhance.c#L1551

> 4.Why?

See above.

**Reply by EricBatlle** [reply]
> > How can I work with memory images (byte[]) instead of static ones? I expected that this library with the new ImageMagcik(byte[]) can do that, but I'm stuck trying to find how to apply convert to those virtual images...
> 
> I don't understand what you are asking me. If you want to read a single image you can specify the bytes of the image or the name of the file (and some other methods) as an argument to the `MagickImage` class.
> 
> > There is a way with Magick.NET to achieve that?
> 
> Still not sure what you are asking?
> 
> > Why takes so long to perform a -normalize filter?
> 
> The duration also depends on the version of ImageMagick that you are using. You are probably using a `Q16-HDRI` version on the command line. You might want to consider to use the `Q8` version of Magick.NET instead. And another thing that might help would be to use the `OpenMP` version.
> 


Thanks dlmestra for your feedback, but clearly I wasn't clear about my doubts, I'll try again:

1. Does Magick.NET have the ImageMagick "convert" command? If yes, how can I used? Cause I can't find it on Magick.NET repo documentation.
2. The benchmarks that I mention weren't taken with Magick.NET, but with ImageMagick commands (cause as Explained in 1, I don't know how to call convert -normalize from Magick.NET)
3. In case Magick.NET can't perform the convert -normalize command, it is possible to use ImageMagick, from the command line, using byte[] instead of the stored in disk image? How?

> And I suspect the duration is longer because we use double precision instead. More info about the implementation can be found here:
> 
> https://github.com/ImageMagick/ImageMagick/blob/8c4418952feafa9e5699404005d528a838db046a/MagickCore/enhance.c#L4119
> 
> And here:
> 
> https://github.com/ImageMagick/ImageMagick/blob/8c4418952feafa9e5699404005d528a838db046a/MagickCore/enhance.c#L1551
> 

Many many thanks for those links, hope this time my questions are better defined :)

**Reply by dlemstra** [reply]
> 1. Does Magick.NET have the ImageMagick "convert" command? If yes, how can I used? Cause I can't find it on Magick.NET repo documentation.

It has (almost) all the functionality that is available in the `convert` executable but it uses a C# api instead.

> 2. The benchmarks that I mention weren't taken with Magick.NET, but with ImageMagick commands (cause as Explained in 1, I don't know how to call convert -normalize from Magick.NET)

The `MagickImage` class has `Normalize()` method that does the same as the `-normalize` option.

**Reply by EricBatlle** [reply]
My god, simply as `Normalize()` that was what I was looking for, typingi "convert.." expecting to autocomplete with some argument or something else...my bad, thanks for your patience dlmestra.
